contract-examples/contracts/ERC20NativeMinter.sol

```diff
@@ -56,3 +58,147 @@ contract ERC20NativeMinter is ERC20, AllowList {
     return 18;
   }
 }
+
+// TODO:
+// this contract adds another (unwanted) layer of indirection
+// but it's the easiest way to match the previous HardHat testing functionality.
+// Once we completely migrate to DS-test, we can simplify this set of tests.
+contract Minter {
+  ERC20NativeMinter token;
+
+  constructor(address tokenAddress) {
+    token = ERC20NativeMinter(tokenAddress);
+  }
+
+  function mintdraw(uint amount) external {
+    token.mintdraw(amount);
+  }
+
+  function deposit(uint value) external {
+    token.deposit{value: value}();
+  }
+}
+
+contract ERC20NativeMinterTest is DSTest {
+  function setUp() public {
+    // noop
+  }
+
+  function test_mintdrawFailure() public {
+    ERC20NativeMinter token = new ERC20NativeMinter(1000);
+    address tokenAddress = address(token);
+
+    INativeMinter nativeMinter = INativeMinter(MINTER_ADDRESS);
+
+    assertEq(nativeMinter.readAllowList(tokenAddress), 0);
+
+    try token.mintdraw(100) {
+      assertTrue(false, "mintdraw should fail");
+    } catch {} // TODO should match on an error to make sure that this is failing in the way that's expected
+  }
+
+  function test_addMinter() public {
+    ERC20NativeMinter token = new ERC20NativeMinter(1000);
+    address tokenAddress = address(token);
+
+    INativeMinter nativeMinter = INativeMinter(MINTER_ADDRESS);
+
+    assertEq(nativeMinter.readAllowList(tokenAddress), 0);
+
+    nativeMinter.setEnabled(tokenAddress);
+
+    assertEq(nativeMinter.readAllowList(tokenAddress), 1);
+  }
+
+  function test_adminMintdraw() public {
+    ERC20NativeMinter token = new ERC20NativeMinter(1000);
+    address tokenAddress = address(token);
+
+    address testAddress = address(this);
+
+    INativeMinter nativeMinter = INativeMinter(MINTER_ADDRESS);
+    nativeMinter.setEnabled(tokenAddress);
+
+    uint initialTokenBalance = token.balanceOf(testAddress);
+    uint initialNativeBalance = testAddress.balance;
+
+    uint amount = 100;
+
+    token.mintdraw(amount);
+
+    assertEq(token.balanceOf(testAddress), initialTokenBalance - amount);
+    assertEq(testAddress.balance, initialNativeBalance + amount);
```

Comment:
Please give me a specific example of the level of detail you think is required to understand a test like this. What specifically don't you understand?

Reply:
We can have some inlined comments like here `ERC20 token balance should be decreased` etc. Also comments above setup/teardown functions can be helpful to understand which functions are testing and which are doing auxiliary work.

Reply:
but it should be abundantly clear that the token balance should be decreased.
```solidity
assertEq(token.balanceOf(testAddress), initialTokenBalance - amount);
```
If you think this would be a little clearer
```solidity
unit currentTokenBalance = token.balanceOf(testAddress);
assertEq(currentTokenBalance, intialTokenBalance - amount); 
```
I'm happy to make that change. 

Excessive commenting is usually a symptom of unclear code that can likely be improved by actual code changes (better naming or otherwise) rather than adding comments.